test/e2e/data/ci-artifacts-platform-kustomization.yaml

```diff
@@ -0,0 +1,336 @@
+# Can be used to modify the resources when use-ci-artifacts is enabled.
+---
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+kind: KubeadmControlPlane
+metadata:
+  name: ${CLUSTER_NAME}-control-plane
+spec:
+  infrastructureTemplate: {}
+  kubeadmConfigSpec:
+    files:
+      # We need to add these files again, because kustomize overrides the whole array
+      # See: https://github.com/kubernetes-sigs/kustomize/issues/2825
+      - content: ${OPENSTACK_CLOUD_PROVIDER_CONF_B64}
+        encoding: base64
+        owner: root
+        path: /etc/kubernetes/cloud.conf
+        permissions: "0600"
+      - content: ${OPENSTACK_CLOUD_CACERT_B64}
+        encoding: base64
+        owner: root
+        path: /etc/certs/cacert
+        permissions: "0600"
+      - content: |
+          #!/bin/bash
+
+          ## Please note that this file needs to be escaped for envsubst to function
+
+          function retry {
+            attempt=0
+            max_attempts=$${1}
+            interval=$${2}
+            shift; shift
+            until [[ $${attempt} -ge "$${max_attempts}" ]] ; do
+              attempt=$((attempt+1))
+              set +e
+              eval "$*" && return || echo "failed $${attempt} times: $*"
+              set -e
+              sleep "$${interval}"
+            done
+            echo "error: reached max attempts at retry($*)"
+            return 1
+          }
+          retry 10 5 "apt-get update"
+
+        owner: root:root
+        path: /usr/local/bin/retry-apt.sh
+        permissions: "0750"
+      - content: |
+          #!/bin/bash
+
+          # Copyright 2020 The Kubernetes Authors.
+          #
+          # Licensed under the Apache License, Version 2.0 (the "License");
+          # you may not use this file except in compliance with the License.
+          # You may obtain a copy of the License at
+          #
+          #     http://www.apache.org/licenses/LICENSE-2.0
+          #
+          # Unless required by applicable law or agreed to in writing, software
+          # distributed under the License is distributed on an "AS IS" BASIS,
+          # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+          # See the License for the specific language governing permissions and
+          # limitations under the License.
```

Comment:
This is noisy.. I am not familiar with law, but do we really need this?

Reply:
The copyright header?

I would keep it as is for now. But, I wanted to do a follow-up anyway to get rid of the whole `/usr/local/bin/ci-artifacts.sh` file in this YAML anyway. There are some reasons why we are not using the upstream function to add this functionality:
1. they are using gsutil instead of  curl. Thus they install google cloud sdk first which adds  ~3-5 minutes to to node startup in our environment (probably in all environments)
2. they are also pulling the "kube-apiserver" "kube-controller-manager" "kube-scheduler" for worker nodes, which is not necessary and also adds time to the test 
3/ when we use the upstream method to patch our KubeadmControlPlane and KubeadmConfigTemplate to add the ` `/usr/local/bin/ci-artifacts.sh` script we will lose the `/etc/kubernetes/cloud.conf` and `/etc/certs/cacert` files because the whole `files` array is overwritten. This is a limitation of kustomize, but I assume with the fixes merged to kustomize a week ago we are able to improve this (https://github.com/kubernetes-sigs/kustomize/issues/2825) (although it's not completely trivial)

I wanted to do a follow-up to improve all those things in the `cluster-api` util func `GenerateCIArtifactsInjectedTemplateForDebian` I currently copied into `tmp_template.go` and then drop this in our repo. I think until know I would like to keep the diff as small as possible.

Reply:
I opened a issue for it here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/804

Reply:
Thanks, understand.